In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
import warnings
warnings.filterwarnings('ignore')
df = pd.read_csv("data/sonar_dataset.csv", header=None)
df.sample(5)

,0,1,2,3,4,5,6,7,8,9,...,51,52,53,54,55,56,57,58,59,60
48,0.0190,0.0038,0.0642,0.0452,0.0333,0.0690,0.0901,0.1454,0.0740,0.0349,...,0.0093,0.0118,0.0112,0.0094,0.0140,0.0072,0.0022,0.0055,0.0122,R
42,0.0211,0.0319,0.0415,0.0286,0.0121,0.0438,0.1299,0.1390,0.0695,0.0568,...,0.0090,0.0042,0.0153,0.0106,0.0020,0.0105,0.0049,0.0070,0.0080,R
186,0.0209,0.0191,0.0411,0.0321,0.0698,0.1579,0.1438,0.1402,0.3048,0.3914,...,0.0078,0.0201,0.0104,0.0039,0.0031,0.0062,0.0087,0.0070,0.0042,M
40,0.0068,0.0232,0.0513,0.0444,0.0249,0.0637,0.0422,0.1130,0.1911,0.2475,...,0.0173,0.0163,0.0055,0.0045,0.0068,0.0041,0.0052,0.0194,0.0105,R
30,0.0240,0.0218,0.0324,0.0569,0.0330,0.0513,0.0897,0.0713,0.0569,0.0389,...,0.0162,0.0146,0.0093,0.0112,0.0094,0.0054,0.0019,0.0066,0.0023,R


In [4]:
df.shape

(208, 61)

In [5]:
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [6]:
df.columns

Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53,
       54, 55, 56, 57, 58, 59, 60],
      dtype='int64')

In [7]:
df[60].value_counts()

60
M    111
R     97
Name: count, dtype: int64

In [8]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [9]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5)

,R
100,False
3,True
56,True
40,True
178,False


In [10]:
y.value_counts()

R    
False    111
True      97
Name: count, dtype: int64

In [11]:
X.head()

,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
0,0.0200,0.0371,0.0428,0.0207,0.0954,0.0986,0.1539,0.1601,0.3109,0.2111,...,0.0232,0.0027,0.0065,0.0159,0.0072,0.0167,0.0180,0.0084,0.0090,0.0032
1,0.0453,0.0523,0.0843,0.0689,0.1183,0.2583,0.2156,0.3481,0.3337,0.2872,...,0.0125,0.0084,0.0089,0.0048,0.0094,0.0191,0.0140,0.0049,0.0052,0.0044
2,0.0262,0.0582,0.1099,0.1083,0.0974,0.2280,0.2431,0.3771,0.5598,0.6194,...,0.0033,0.0232,0.0166,0.0095,0.0180,0.0244,0.0316,0.0164,0.0095,0.0078
3,0.0100,0.0171,0.0623,0.0205,0.0205,0.0368,0.1098,0.1276,0.0598,0.1264,...,0.0241,0.0121,0.0036,0.0150,0.0085,0.0073,0.0050,0.0044,0.0040,0.0117
4,0.0762,0.0666,0.0481,0.0394,0.0590,0.0649,0.1209,0.2467,0.3564,0.4459,...,0.0156,0.0031,0.0054,0.0105,0.0110,0.0015,0.0072,0.0048,0.0107,0.0094


In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)
X_train.head()


,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
67,0.0368,0.0403,0.0317,0.0293,0.0820,0.1342,0.1161,0.0663,0.0155,0.0506,...,0.0058,0.0091,0.0160,0.0160,0.0081,0.0070,0.0135,0.0067,0.0078,0.0068
14,0.0124,0.0433,0.0604,0.0449,0.0597,0.0355,0.0531,0.0343,0.1052,0.2120,...,0.0078,0.0083,0.0057,0.0174,0.0188,0.0054,0.0114,0.0196,0.0147,0.0062
164,0.0163,0.0198,0.0202,0.0386,0.0752,0.1444,0.1487,0.1484,0.2442,0.2822,...,0.0027,0.0077,0.0026,0.0031,0.0083,0.0020,0.0084,0.0108,0.0083,0.0033
179,0.0394,0.0420,0.0446,0.0551,0.0597,0.1416,0.0956,0.0802,0.1618,0.2558,...,0.0118,0.0146,0.0040,0.0114,0.0032,0.0062,0.0101,0.0068,0.0053,0.0087
19,0.0126,0.0149,0.0641,0.1732,0.2565,0.2559,0.2947,0.4110,0.4983,0.5920,...,0.0153,0.0092,0.0035,0.0098,0.0121,0.0006,0.0181,0.0094,0.0116,0.0063


Model without Dropout Layer

In [13]:
import tensorflow as tf
from tensorflow import keras
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 1s 777us/step - accuracy: 0.5000 - loss: 0.7098 
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 608us/step - accuracy: 0.5385 - loss: 0.6882
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 733us/step - accuracy: 0.5577 - loss: 0.6797
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 695us/step - accuracy: 0.5769 - loss: 0.6707
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 705us/step - accuracy: 0.6090 - loss: 0.6552
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 631us/step - accuracy: 0.6474 - loss: 0.6336
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 627us/step - accuracy: 0.6795 - loss: 0.6112
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 621us/step - accuracy: 0.6923 - loss: 0.5850
Epoch 9/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 551us/step - accuracy: 0.7756 - loss: 0.5654
Epoch 10/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 552us/step - accuracy: 0.8077 - loss: 0.5291
Epoch 11/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 524us/step - accuracy: 0.8269 - loss: 0.4904
Epoch 12/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 

In [14]:
model.evaluate(X_test, y_test)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7885 - loss: 0.9374 


[0.9373968839645386, 0.7884615659713745]

In [15]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step
[5.8844665e-11 9.5618337e-01 9.6218538e-01 4.6059213e-06 9.9990135e-01
 9.9951810e-01 2.0149987e-02 9.9993271e-01 9.2326866e-07 9.9994326e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [16]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.74      0.93      0.82        27
        True       0.89      0.64      0.74        25

    accuracy                           0.79        52
   macro avg       0.81      0.78      0.78        52
weighted avg       0.81      0.79      0.78        52



Model with Dropout Layer

In [17]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 1s 649us/step - accuracy: 0.5385 - loss: 0.7191 
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 636us/step - accuracy: 0.5064 - loss: 0.7069
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 681us/step - accuracy: 0.5256 - loss: 0.6963
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 651us/step - accuracy: 0.5449 - loss: 0.6937
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 634us/step - accuracy: 0.5385 - loss: 0.6821
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 595us/step - accuracy: 0.5256 - loss: 0.7081
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 599us/step - accuracy: 0.5577 - loss: 0.6863
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 582us/step - accuracy: 0.6218 - loss: 0.6861
Epoch 9/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 537us/step - accuracy: 0.5577 - loss: 0.6915
Epoch 10/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 530us/step - accuracy: 0.6346 - loss: 0.6718
Epoch 11/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 518us/step - accuracy: 0.5513 - loss: 0.6829
Epoch 12/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 

In [18]:
modeld.evaluate(X_test, y_test)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7692 - loss: 0.4588 


[0.45879533886909485, 0.7692307829856873]

In [19]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 24ms/step 
[0.00390013 0.8775763  0.9357425  0.04030248 0.99987376 0.937992
 0.6120918  0.99984163 0.03659782 0.99997026]
[0. 1. 1. 0. 1. 1. 1. 1. 0. 1.]


In [20]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.74      0.85      0.79        27
        True       0.81      0.68      0.74        25

    accuracy                           0.77        52
   macro avg       0.78      0.77      0.77        52
weighted avg       0.77      0.77      0.77        52



You can see that by using dropout layer test accuracy increased from 0.77 to 0.81